### PR TITLE
Allow commenting on pipeline instance

### PR DESCRIPTION
**Pull request by bdpiparva**
Issue: #7366 

comment-on-pipeline-run

### CODE REVIEW

File: spark/spark-spa/src/main/java/com/thoughtworks/go/spark/spa/PipelineActivityController.java [hunk 1]
```diff
@@ -49,7 +49,7 @@ public String controllerBasePath() {
     @Override
     public void setupRoutes() {
         path(controllerBasePath(), () -> {
-            before("", authenticationHelper::checkUserAnd403);
+            before("", authenticationHelper::checkPipelineViewPermissionsAnd403);
```
Comment (kritika-singh3) [inline spark/spark-spa/src/main/java/com/thoughtworks/go/spark/spa/PipelineActivityController.java]: No tests changed?

File: server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/pipeline_run_info_widget.tsx [hunk 2]
```diff
@@ -77,6 +79,9 @@ export class PipelineRunWidget extends MithrilViewComponent<PipelineRunAttrs> {
             </div>
           </div>;
         })}
+        <CommentWidget comment={pipelineRunInfo.comment}
+                       counterOrLabel={pipelineRunInfo.counterOrLabel()}
+                       addOrUpdateComment={vnode.attrs.addOrUpdateComment}/>
```
Comment (kritika-singh3) [inline server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/pipeline_run_info_widget.tsx]: The page is shown when the user has `view` permissions. But the comment feature is available only when `operate` permission is provided. Should we check that before showing the button to add?

File: server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/comment_widget.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {MithrilViewComponent} from "jsx/mithril-component";
+import _ from "lodash";
+import m from "mithril";
+import Stream from "mithril/stream";
+import {Secondary} from "../../components/buttons";
```
Comment (kritika-singh3) [inline server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/comment_widget.tsx]: Suggestion: We prefer absolute import path `views/components/...` over relative one. Please update this.

File: server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/pipeline_run_info_widget.tsx (hunk 2 above)
Comment (bdpiparva) [reply server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/pipeline_run_info_widget.tsx]: Yes, Ideally it should be visible only when the user has `operate` permission. There is no point in showing the button as an API call will always fail.

Comment (kritika-singh3) [reply server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/pipeline_run_info_widget.tsx]: Also, pls check the following scenario:
1. `user1` adds a comment.
2. Another `user2` can view it but cannot edit or remove it.

File: spark/spark-spa/src/main/java/com/thoughtworks/go/spark/spa/PipelineActivityController.java (hunk 1 above)
Comment (bdpiparva) [reply spark/spark-spa/src/main/java/com/thoughtworks/go/spark/spa/PipelineActivityController.java]: Done

File: server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/pipeline_run_info_widget.tsx (hunk 2 above)
Comment (bdpiparva) [reply server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/pipeline_run_info_widget.tsx]: @kritika-singh3  
- Currently, anyone can add/edit the comment on the pipeline instance if it has `operate` permission on the pipeline(same as old SPA).

- Adding comment will fail if the user don't have the `operate` permission on the pipeline

File: server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/comment_widget.tsx (hunk 3 above)
Comment (bdpiparva) [reply server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipeline_activity/comment_widget.tsx]: Done

**Pull request merged**